curta/src/chip/hash/sha/keccak256/mod.rs

```diff
@@ -0,0 +1,166 @@
+        let c_arr = self.alloc_array::<U64Register>(5);
+        let d_arr = self.alloc_array::<U64Register>(5);
+        for x in 0..5 {
+            let mut c_i = state.get(x); 
+            for y in 1..5 {
+                c_i = self.bitwise_xor(&c_i, &state.get(y * 5 + x), operations);
+            }
+            // Does it suffice to constrain for every row of c_i, it need to satisfy its relationship with state at the same row?
+            self.assert_equal_transition(&c_arr.get(x), &c_i);
+        }
+        // initial state doesn't exist? or don't want to constrain? first row of it is empty?
+        let state_after_theta = self.alloc_array::<U64Register>(25);
+        
+        for x in 0..5 {
+            let temp = self.bit_rotate_right(&c_arr.get((x + 1) % 5), 1, operations);
+            let d_i = self.bitwise_xor(&c_arr.get((x+4)%5), &temp, operations);
+            self.assert_equal_transition(&d_arr.get(x), &d_i);
+            for y in 0..5 {
+                // make sure next row of state_after_theta follows the theta transition of state
+                let tmp = self.bitwise_xor(&state.get(y * 5 + x), &d_i, operations);
+                self.set_to_expression_transition(&(state_after_theta.get(y * 5 + x)).next(), tmp.expr());
+            }
+        }
+
+        let state_after_rhopi = self.alloc_array::<U64Register>(25);
+
+        // rho and pi
+        for x in 0..5 {
+            for y in 0..5 {
+                // x is column, y is row
+                // y, 2x+3y, is pi_idx in the flatten version
+                let pi_idx = ((2 * x + 3 * y) % 5) * 5 + y;
+                let tmp = self.bit_rotate_right(&state_after_theta.get(y * 5 + x), KECCAKF_ROTC[y][x].try_into().unwrap(), operations);
+                self.set_to_expression_transition(&state_after_rhopi.get(pi_idx).next(), tmp.expr());
+            }
+        }
+
+        let state_after_chi = self.alloc_array::<U64Register>(25);
+        // chi
+        for x in 0..5 {
+            for y in 0..5 {
+              
+                let tmp1 = self.bitwise_not(&state_after_rhopi.get(  (x + 1) % 5 + y * 5), operations);
+                let tmp2 = self.bitwise_and(&tmp1, &state_after_rhopi.get((x+2) % 5 + y * 5), operations);
+                let tmp3 = self.bitwise_xor(&state_after_rhopi.get(x + y * 5), &tmp2, operations);
+                self.set_to_expression_transition(&state_after_chi.get(x + y*5), tmp3.expr());
+            }
+        }
+
+        // iota
+        for i in 0..24 {
+            let tmp = self.bitwise_xor(&state_after_chi.get(0), &round_const.get(i), operations);
+            self.set_to_expression_transition(&state.get(0).next(),
+            tmp.expr());
+        }
+
+        // constrain the other val of next state is the same as chi's result
+        for x in 0..5 {
+            for y in 0..5 {
+                // can if used here? if not, guess need to pass some indicator bit to help identity 0,0 
```

Comment:
yes, it can. remember that the if is used to define constraints here which hold on all rows, so it is a compile time thing